### ISSUE TITLE
Support EIP-712 signing for TREZOR accounts

### ISSUE THREAD

**Issue by angyts** [MetaMask/metamask-extension#11498]
Dear Devs, thank you for reading this. 

This is a long standing problem that breaks support of hardware wallets for many web 3 applications. 

I know that ledger already currently supports with a [not ideal solution](https://github.com/MetaMask/metamask-extension/issues/10240) but working nonetheless.

Is anyone looking into the trezor side of things?

The devs on the trezor side are currently looking into the following to fully support EIP-712.
https://github.com/trezor/trezor-firmware/pull/1568

It would be lovely if Metamask could look into the `connect` part of the specifications. I'm sorry I cannot help as this is really not in my field of expertise.

**Comment by brad-decker** [comment]
Could you say more about why the Ledger solution is not ideal?

**Comment by angyts** [comment]
Please correct me if I'm wrong.

I think the ledger team gets the user sign the hash of the message, whereas the trezor team is trying to port the entire message over. 

I personally would prefer to see the entire message i'm signing.

**Comment by darkwing** [comment]
There is presently a PR open for this (https://github.com/trezor/trezor-firmware/pull/1568) which Trezor must merge before we can complete this task.

**Comment by angyts** [comment]
You are absolutely right about that. 

Thank you for looking into in. So that when the trezor side has merged, metamask will be able to support.

**Comment by Gudahtt** [comment]
>There is presently a PR open for this (trezor/trezor-firmware#1568) which Trezor must merge before we can complete this task.

It looks like that PR was superseded by https://github.com/trezor/trezor-firmware/pull/1835, which has been merged! :tada: 

I don't know whether it has been released yet or not. Once it has been released, this issue will no longer be blocked.

**Comment by yura-pakhuchiy** [comment]
Firmware 2.4.3 for Trezor T with EIP-712 is available now.
https://github.com/trezor/webwallet-data/pull/32/files
@darkwing @danjm

**Comment by gewure** [comment]
Any Update on this?

**Comment by blueblack2** [comment]
Does the EIP 712 firmware update for the Model T work with Metamask yet?

**Comment by blueblack2** [comment]
> [<img alt="@blueblack2" width="40" height="40" src="https://avatars.githubusercontent.com/u/84169276?s=80&amp;v=4">](/blueblack2)

Does this Firmware EIP 712 update for the Model T work with signing transactions with Metamask yet?

**Comment by blueblack2** [comment]
> The new Trezor T firmware has been released, supporting EIP-712
> 
> Metamask is now un-blocked on this issue, hopefully the feature will be supported soon.

Does this Firmware EIP 712 update for the Model T work with signing transactions with Metamask yet?

**Comment by astanciu** [comment]
Any updates here?  As soon as you try to sign anything, the following error shows up in console:
`MetaMask Message Signature: Error: Not supported on this device`

This is using Trezor T, 2.4.3, MM@10.8.0

**Comment by aloisklink** [comment]
Just to give people an update, I've added support to `trezor-connect` (Trezor JavaScript library) for the Trezor Model T in https://github.com/trezor/connect/pull/983.

We're currently waiting for the next release of `trezor-connect` before I work on adding it to the MetaMask side. I'm not an employee of @MetaMask or @TREZOR and can't promise anything, but depending on how much free time I have, I should be able to add it pretty quickly once it releases, since I'm already familiar with the code.

Trezor One support has already been added to the firmware, and should arrive in the next firmware release. `trezor-connect` EIP-712 support for T1 is also being worked on in https://github.com/trezor/connect/pull/986.

**Comment by RudyB** [comment]
Trezor added support for EIP-712 for the Model One as of January 19, 2022 however it remains unsupported in Metamask

**Comment by astanciu** [comment]
For anyone following, this is the piece we're waiting on. It's done and can be manually installed, just not shipped
https://github.com/MetaMask/eth-trezor-keyring/pull/117#issuecomment-1040733651

**Comment by msubash26** [comment]
Hi Alex, any timeline on by when this will be shipped? many users are stuck with Trezor model one after the recent Wyvern exchange contract upgrade by Opensea, Please address this as early as possible. Thanks for the support!

**Comment by astanciu** [comment]
> Hi Alex, any timeline on by when this will be shipped? many users are stuck with Trezor model one after the recent Wyvern exchange contract upgrade by Opensea, Please address this as early as possible. Thanks for the support!


No clue, I'm not involved with the project, was just sharing the link ;)

**Comment by metaver5o** [comment]
still not working!

**Comment by blueblack2** [comment]
Is this working yet?

**Comment by digitalgermination** [comment]
I can attest that this does not appear to have resolved the issue. Still can't use Trezor on OpenSea or LooksRare

**Comment by FrederikBolding** [comment]
The feature was merged but has not been released yet. Will possibly be included in the next release!

**Comment by digitalgermination** [comment]
> The feature was merged but has not been released yet. Will possibly be included in the next release!

Oh, thank you for the heads up!

**Comment by KOOLMOLL** [comment]
В правильности выбора вопрос?и без инструментов влияния алгоритм тот же.

**Comment by blueblack2** [comment]
When will this be released?

**Comment by msubash26** [comment]
Why the latest release ([Version 10.10.1] has the fix for Lattice hardware wallet support to sign EIP-712 and not for Trezor though its fixed as per the comments above?

**Comment by astanciu** [comment]
Yeah seriously, why wasn't this included?

**Comment by isaiahcurry8** [comment]
Lattice but not Trezor? You guys are stupid.

**Comment by FrederikBolding** [comment]
@msubash26 @astanciu @isaiahcurry8 It has been included in the release candidate for: `10.11.0`. Which should be the next release, but it needs to go through the QA process first!

For `10.10.1` we fixed an issue with the existing Lattice EIP.712 integration, which was already released and tested, thus it could go out faster.

**Comment by msubash26** [comment]
@FrederikBolding , Thanks for the update, could you confirm the release date for 10.11.0?

**Comment by ching2302** [comment]
> @msubash26 @astanciu @spoonknife22 It has been included in the release candidate for: `10.11.0`. Which should be the next release, but it needs to go through the QA process first!
> 
> For `10.10.1` we fixed an issue with the existing Lattice EIP.712 integration, which was already released and tested, thus it could go out faster.

Hi Sir, any idea regarding 10.11.0 release date?

**Comment by blueblack2** [comment]
Is this working yet? According to this https://github.com/MetaMask/metamask-extension/releases
10.11 hasn't included support for Trezor EIP 712 signing, when will this be released?

**Comment by blueblack2** [comment]
> @msubash26 @astanciu @spoonknife22 It has been included in the release candidate for: `10.11.0`. Which should be the next release, but it needs to go through the QA process first!
> 
> For `10.10.1` we fixed an issue with the existing Lattice EIP.712 integration, which was already released and tested, thus it could go out faster.

Is this working yet? According to this https://github.com/MetaMask/metamask-extension/releases
10.11 hasn't included support for Trezor EIP 712 signing, when will this be released?

**Comment by FrederikBolding** [comment]
@blueblack2 It was released in `10.10.2` instead and should be live as of last week!
